### PR TITLE
Laboratory work #2, Maria Kramkova - 22FPL1

### PR DESCRIPTION
**Pull request by kramkova**

### CODE REVIEW

File: lab_2_tokenize_by_bpe/main.py [hunk 1]
```diff
@@ -36,6 +57,15 @@ def count_tokens_pairs(
...
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+    pair_dict = {}
+    for word in word_frequencies:
+        for i in range(len(word) - 1):
+            if not (word[i], word[i + 1]) in pair_dict:
+                pair_dict[(word[i], word[i + 1])] = 0
+            pair_dict[(word[i], word[i + 1])] += 1 * word_frequencies[word]
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:67]: better multiply to `0`

File: lab_2_tokenize_by_bpe/main.py [hunk 2]
```diff
@@ -47,6 +77,31 @@ def merge_tokens(
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if (not isinstance(word_frequencies, dict)
+       or not isinstance(pair, tuple)):
+        return None
+    new_dict = {}
+    for word in word_frequencies:
+        new_word = []
+        ''' Alternative variation with a switch:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:86]: the actual and alternative implementations are strange. Wee need the third

File: lab_2_tokenize_by_bpe/main.py [hunk 3]
```diff
@@ -69,6 +137,21 @@ def get_vocabulary(
...
+    tokens = [unknown_token]
+    for word in word_frequencies:
+        for word_part in word:
+            tokens.append(word_part)
+            for symbol in word_part:
+                if symbol not in tokens:
+                    tokens.append(symbol)
+    tokens.sort(key=lambda x: (-len(x), str(x[0])))
+    id_clues = {}
+    for i in range(len(tokens)):
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:152]: enumerate

File: lab_2_tokenize_by_bpe/main.py [hunk 4]
```diff
@@ -69,6 +137,21 @@ def get_vocabulary(
     :param unknown_token: a token to signify an unknown token
     :return: dictionary in the form of <token: identifier>
     """
+    if (not isinstance(word_frequencies, dict)
+            or not isinstance(unknown_token, str)):
+        return None
+    tokens = [unknown_token]
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:143]: why not a `set`?

File: lab_2_tokenize_by_bpe/main.py [hunk 5]
```diff
@@ -69,6 +137,21 @@ def get_vocabulary(
...
+    for word in word_frequencies:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:144]: too many `for`s

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Where are changes in `start.py`?

File: lab_2_tokenize_by_bpe/main.py [hunk 6]
```diff
@@ -36,6 +57,15 @@ def count_tokens_pairs(
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+    pair_dict = {}
+    for word in word_frequencies:
+        for i in range(len(word) - 1):
+            if not (word[i], word[i + 1]) in pair_dict:
```
[the remark this answers is not in the document]
Comment (kramkova) [reply lab_2_tokenize_by_bpe/main.py:65]: done!

File: lab_2_tokenize_by_bpe/main.py (hunk 5 above)
Comment (kramkova) [reply lab_2_tokenize_by_bpe/main.py:144]: now there are two of them

File: lab_2_tokenize_by_bpe/main.py (hunk 3 above)
Comment (kramkova) [reply lab_2_tokenize_by_bpe/main.py:152]: done

File: lab_2_tokenize_by_bpe/main.py (hunk 4 above)
Comment (kramkova) [reply lab_2_tokenize_by_bpe/main.py:143]: now set() helps shrink the list after finding all possible tokens

File: lab_2_tokenize_by_bpe/main.py (hunk 1 above)
Comment (kramkova) [reply lab_2_tokenize_by_bpe/main.py]: simplified this function a bit in hopes that it is what you meant

File: lab_2_tokenize_by_bpe/main.py (hunk 2 above)
Comment (kramkova) [reply lab_2_tokenize_by_bpe/main.py]: the third has been introduced

File: lab_2_tokenize_by_bpe/main.py [hunk 7]
```diff
@@ -47,6 +81,21 @@ def merge_tokens(
...
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if (not isinstance(word_frequencies, dict)
+       or not isinstance(pair, tuple)):
+        return None
+    new_dict = {}
+    merged_pair = ''.join(pair)
+    for word, freq in word_frequencies.items():
+        merging_word = list(word)
+        for i, token in enumerate(merging_word[:-1]):
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:91]: try to optimize it

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**